File: src/hw7/Boggle.clj
; Hunter Damron
; Copyright 2019
;
; NOTE: Submitting any part of this assignment for credit is plagiarism.
;   Referencing this material to any extent without citation is plagiarism.

(use '[clojure.java.io :only (reader)])
(use '[clojure.string :only (split upper-case join)])
(import java.io.BufferedReader)

; Size of board
(def N 5)
(def DICTFILE "data/hw7/dict.txt")

; Load dictionary from file
(def dict (map upper-case (with-open [scan (reader DICTFILE)] (reduce conj [] (line-seq scan)))))

; Load board from stdin as 2d array of chars
(def board (to-array-2d (map #(map first (split % #" " N)) (take N (line-seq (BufferedReader. *in*))))))

; Gets indices +/- 1 from i
(defn neighbors-1d [i] (map (partial + i) (range -1 2)))

; Gets 2d indices +/- 1 from i,j
(defn neighbors-2d [i j] (for [ni (neighbors-1d i) nj (neighbors-1d j)] [ni nj]))

; Gets all 2d indices around i j which are in bounds and not already used
(defn neighbors [i j N used] (filter
  #(apply (fn [ni nj] (and
    (or (not= ni i) (not= nj j))
    (>= ni 0) (>= nj 0) (< ni N) (< nj N)
    (not (contains? used [ni nj])))) %)
  (neighbors-2d i j)))

; Gets words in dict which have character c at index i
(defn subdict [dict i c] (filter #(= (get % i) c) dict))

; Searches board recursively for boggle words in dict
(defn search-board [prev used i j dict board]
  (map #(apply (fn [ni nj]
    (let
      [ c (aget board ni nj)
        w (join [prev (str c)])
        subd (subdict dict (count prev) c)
        keeper (and (not (empty? subd)) (= (first subd) w))
        psubd (if keeper (rest subd) subd)
        sub-matches (if-not (empty? psubd) (search-board w (conj used [ni nj]) ni nj psubd board) []) ]
      (keep identity (conj sub-matches (when keeper (first subd)))))) %)
    (neighbors i j N used)))

; Wrapper around search-board for a single start position i j
(defn search-from-ij [i j dict board]
  (let [c (aget board i j)] (flatten (search-board (str c) #{[i j]} i j (subdict dict 0 c) board))))

; Loops over start positions and writes the found words from each
(doseq [i (range N) j (range N)]
  (let [matches (map #(join ["Found Word: " %]) (search-from-ij i j dict board))]
    (println (format "Starting %d %d" i j))
    (when-not (empty? matches) (println (join \newline matches)))))

File: src/hw7/Boggle.hs
-- Hunter Damron
-- Copyright 2019
--
-- NOTE: Submitting any part of this assignment for credit is plagiarism.
--   Referencing this material to any extent without citation is plagiarism.

import System.IO
import Data.Char
import Data.Set (Set, singleton, notMember, insert)

dictFile = "data/hw7/dict.txt"

loadDict :: IO [String]
loadDict = do
  contents <- readFile dictFile
  return . lines . (map toUpper) $ contents

loadBoard :: IO [[Char]]
loadBoard = do
  contents <- getContents
  return . (map $ (map head) . words) . lines $ contents

get2d :: (Int, Int) -> [[a]] -> a
get2d (i, j) l = l !! i !! j

inBounds :: (Int, Int) -> Int -> Set (Int, Int) -> Bool
inBounds (i, j) n used = i >= 0 && j >= 0 && i < n && j < n && (notMember (i, j) used)

neighbors :: (Int, Int) -> Int -> Set (Int, Int) -> [(Int, Int)]
neighbors (i, j) n used = [(ni,nj) | ni <- [i-1..i+1], nj <- [j-1..j+1],
                           ni /= i || nj /= j, inBounds (ni, nj) n used]

subdict :: Int -> Char -> [String] -> [String]
subdict i c = filter (\w -> length w > i && w !! i == c)

subsearchBoard :: String -> Set (Int, Int) -> [String] -> [[Char]] -> (Int, Int) -> [String]
subsearchBoard prev used dict board nij = let
  c = get2d nij board
  w = prev ++ [c]
  subd = subdict (length used) c dict
  nused = insert nij used
  nempty = not $ null subd
  match = if (nempty && (head subd) == w) then [w] else []
  submatches = if nempty then searchBoard w nused subd board nij else []
  in match ++ submatches

searchBoard :: String -> Set (Int, Int) -> [String] -> [[Char]] -> (Int, Int) -> [String]
searchBoard prev used dict board ij = concat $ map
  (subsearchBoard prev used dict board)
  (neighbors ij (length board) used)

searchFromij :: [String] -> [[Char]] -> (Int, Int) -> [String]
searchFromij dict board ij = let
  c = get2d ij board
  in searchBoard [c] (singleton ij) (subdict 0 c dict) board ij

startingMsg :: (Int, Int) -> String
startingMsg (i,j) = "Starting " ++ (show i) ++ " " ++ (show j)

foundMsg :: String -> String
foundMsg s = "Found Word: " ++ s

strFromij :: [String] -> [[Char]] -> (Int, Int) -> String
strFromij dict board ij = unlines $ [startingMsg ij] ++ (map foundMsg (searchFromij dict board ij))

strSearchBoard :: [String] -> [[Char]] -> String
strSearchBoard dict board = let
  n = length board
  in concat $ map (strFromij dict board) [(i,j) | i <- [0..n-1], j <- [0..n-1]]

main :: IO ()
main = do
  dict <- loadDict
  board <- loadBoard
  putStr $ strSearchBoard dict board
